### PR TITLE
Less write lock contention, better read timeout handling

### PR DESCRIPTION
**Pull request by magik6k**
This PR aims to improve/fix https://github.com/filecoin-project/lotus/issues/8362
* Reduces lock contention on the write lock - it should now be held only when we're actually sending data, response marshaling happens async
* Makes sure that read deadline is reset even if a single frame is taking a long time to read  
* Makes frame unmarshaling not block data reading until we're really, really backlogged
* Adds pprof labels to goroutines handling jsonrpc calls, so if we need to debug this further it's possible to tell apart clients in goroutine dumps.

### CODE REVIEW

File: rpc_test.go [hunk 1]
```diff
@@ -1414,3 +1456,106 @@ func TestReverseCallAliased(t *testing.T) {
...
+	defer cancel()
+
+	ch, err := client.Ch(chctx)
+	require.NoError(t, err)
+
+	prevN := <-ch
+
+	go func() {
+		for n := range ch {
+			fmt.Println("received")
```
Comment (ZenGround0) [inline rpc_test.go:1549]: Does this spam the logs?

File: websocket.go [hunk 2]
```diff
@@ -609,8 +620,63 @@ func (c *wsConn) tryReconnect(ctx context.Context) bool {
...
 }
 
+func (c *wsConn) readFrame(ctx context.Context, r io.Reader) {
+	// debug util - dump all messages to stderr
+	// r = io.TeeReader(r, os.Stderr)
+
+	// json.NewDecoder(r).Decode would read the whole frame as well, so might as well do it
+	// with ReadAll which should be much faster
+	// use a autoResetReader in case the read takes a long time
+	buf, err := io.ReadAll(c.autoResetReader(r)) // todo buffer pool
```
Comment (ZenGround0) [inline websocket.go:630]: Ready to merge without buffer pool?

File: rpc_test.go [hunk 3]
```diff
@@ -1414,3 +1456,106 @@ func TestReverseCallAliased(t *testing.T) {
...
+			if n != prevN+1 {
+				panic("bad order")
```
[the remark this answers is not in the document]
Comment (magik6k) [reply rpc_test.go:1551]: Technically shouldn't use those in a goroutine that isn't running the test, panic was good enough

File: websocket.go (hunk 2 above)
Comment (magik6k) [reply websocket.go:630]: Yeah, that's not really the point of the PR, and can be done separately

File: rpc_test.go (hunk 1 above)
Comment (magik6k) [reply rpc_test.go:1549]: Not that much

File: rpc_test.go [hunk 4]
```diff
@@ -1414,3 +1456,106 @@ func TestReverseCallAliased(t *testing.T) {
...
+			fmt.Println("sending")
+			out <- i
+			i++
+		}
+	}()
+
+	return out, nil
+}
+
+func TestBigResult(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (magik6k) [reply rpc_test.go:1510]: Added some comments

File: websocket.go [hunk 5]
```diff
@@ -609,8 +620,63 @@ func (c *wsConn) tryReconnect(ctx context.Context) bool {
...
+		return
+	}
+
+	c.frameExecQueue <- buf
+	if len(c.frameExecQueue) > cap(c.frameExecQueue)/2 {
+		log.Warnw("frame executor queue is backlogged", "queued", len(c.frameExecQueue), "cap", cap(c.frameExecQueue))
+	}
+
+	// got the whole frame, can start reading the next one in background
+	go c.nextMessage()
```
Comment (Kubuxu) [inline websocket.go:642]: The fact that we are starting goroutines only to kill them and start again bothers me a bit but it shouldn't matter apart from making goroutine numbers annoying to deal with.

File: websocket.go [hunk 6]
```diff
@@ -768,6 +834,36 @@ func (c *wsConn) handleWsConn(ctx context.Context) {
...
+	r     io.Reader
+	reset func()
+
+	lastReset time.Time
+}
+
+func (r *deadlineResetReader) Read(p []byte) (n int, err error) {
+	n, err = r.r.Read(p)
+	if time.Since(r.lastReset) > onReadDeadlineResetInterval {
+		log.Warnw("slow/large read, resetting deadline while reading the frame", "since", time.Since(r.lastReset), "n", n, "err", err, "p", len(p))
```
Comment (Kubuxu) [inline websocket.go:859]: Should this be Warn? It isn't actionable by the user and will happen during any request which requires more than 5s to send.

**Review by Kubuxu — APPROVED** [review]
locking looks correct

File: websocket.go (hunk 6 above)
Comment (magik6k) [reply websocket.go:859]: Arguably whenever any RPC takes that long to transfer, something is broken - so the user should either open an issue, or investigate their networking.

**Pull request merged**